### ISSUE TITLE
3rd Party Plugin Compatibility: Query Monitor

### ISSUE THREAD

**Issue by theperfectwill** [wpsharks/comet-cache#686]
I just came across this... and thought I would mention it. 

I get two errors from Query Monitor. 
- Plugin Reference: https://wordpress.org/plugins/query-monitor/

See screenshot 
- http://take.ms/9gHdZ

These are not reported as "php errors" from Query Monitor, so they don't show up as an issue, unless I read read the query monitor results page. 

But thought I would share as I don't know the validity of the error. It says the method doesn't exist.

**Comment by raamdev** [comment]
@NoahjChampion Are you running ZenCache or Comet Cache? Which version?

**Comment by theperfectwill** [comment]
@raamdev Zencache. 

I know about the change to Comet Cache, and have the version that states that from the Zencache page, but not sure how to update to Comet Cache just yet. I didn't dig into it yet. Do I have to manually download Comet Cache. 

It all happened quickly. I didn't think Comet Cache would be out already :)

**Comment by jaswrks** [comment]
@NoahjChampion Take a look at the "Migrating from ZenCache to Comet Cache" section here. https://cometcache.com/blog/announcing-comet-cache-formerly-zencache/

**Comment by theperfectwill** [comment]
@jaswsinc thanks. Done. 

I took the time to look over all the query monitor results and there are 19 errors, all of the same kind of reference. 

Here is a screenshot of all those errors together @ http://take.ms/sNGqL

**Comment by raamdev** [comment]
I can confirm I'm seeing the same thing with the Query Monitor plugin, however I'm wondering if this is related specifically to how the Query Monitor plugin acquires these results (i.e., is it running each of these hooks itself and then checking the result? and maybe those Comet Cache calls are not meant to be run in that context?).

This will require a bit more research.

Any ideas @jaswsinc?

**Comment by raamdev** [comment]
![2016-02-26_10-04-22](https://cloud.githubusercontent.com/assets/53005/13355842/5778df6c-dc70-11e5-8060-a0382cfb6ddb.png)
![2016-02-26_10-04-07](https://cloud.githubusercontent.com/assets/53005/13355843/58b8d274-dc70-11e5-80a1-1bcc6646f336.png)

**Comment by raamdev** [comment]
After some digging into how Query Monitor works and finding [where it populates the callback handler](https://github.com/johnbillion/query-monitor/blob/2.8.1/classes/Util.php#L147-L220), we were able to narrow this down to an issue with how the Comet Cache codebase is currently doing some tricky things with PHP, which was necessary (until recently) to maintain compatibility with PHP 5.3. 

Now that we have [announced PHP 5.4+](https://cometcache.com/blog/new-minimum-php-version-php-5-4/) is required to run Comet Cache, we will be updating the codebase to convert this "tricky" code to use PHP Traits, which should make the codebase fully compatible with with PHP Reflection class that Query Monitor is using.

---

A few more technical details: 

@jaswsinc writes...

> Yeah, I see what the issue is. They are doing things right, so it's not a bug in their software. This is a PHP limitation.
> 
> On this line: https://github.com/johnbillion/query-monitor/blob/2.8.1/classes/Util.php#L166 they are interpreting an array (e.g., `[$this, 'clearCache']`) as a class Method instead of as a Closure. That's right of them to do that.
> 
> However, this is actually incorrect within Comet Cache, because in reality we are performing some magic in order to work around not being able to use PHP 5.4 code. In our codebase, Closures magically become Methods via `__call()` magic, and so you can't create a ReflectionMethod out of them, because they are defined as Closures and not as class methods. So even though their codebase is checking for both Methods and for Closures, we are able to fool them just enough for it to cause this problem. No easy way around this issue, because they can't predict what magic we might perform in the `__call()` handler.
> 
> What will resolve this is moving to Traits.
> 
> Actually, we don't even use the `__get()` magic in Comet Cache, we are using `$self` with dynamically generated methods, which is even more magical. Magical, in the sense that it can lead to some difficulty when trying to guess where a method comes from with any precision, which is what they are trying to do.

---

So, in summary, once the work in [Issue #635](https://github.com/websharks/comet-cache/issues/635) is complete (this is high on our development list), this compatibility issue with Query Monitor should go away.

**Comment by theperfectwill** [comment]
:+1:

**Comment by raamdev** [comment]
Confirming that the switch to PHP Traits has fixed this issue with Query Monitor (this is a screenshot while running the latest dev build):

![2016-03-09_14-04-58](https://cloud.githubusercontent.com/assets/53005/13646994/0966a054-e600-11e5-9986-12a2330a8f26.png)

**Comment by raamdev** [comment]
Next Release Changelog:
- **Compatibility: Query Monitor.** The Query Monitor plugin was reporting false-positive errors indicating that many Comet Cache methods did not exist. This was due to how the Comet Cache codebase was utilizing PHP Closures, which Query Monitor had a hard time handling. The codebase has been refactored to use PHP Traits instead of Closures and now the Query Monitor plugin has no problem recognizing Comet Cache methods. Props to @NoahjChampion for reporting. See [Issue #686](https://github.com/websharks/comet-cache/issues/686).

**Comment by raamdev** [comment]
**Comet Cache v160416** has been released and includes changes from this GitHub Issue. See the [v160416 announcement](https://cometcache.com/blog/comet-cache-v160416-now-available/) for further details.

---

This issue will now be locked to further updates. If you have something to add related to this GitHub Issue, please open a new GitHub Issue and reference this one (#686).
